Failed tests (10209): TestClientDisconnectLoopback

The following test appears to have failed:

[#10209](https://circleci.com/gh/cockroachdb/cockroach/10209):

```
W1215 22:42:37.215141 564 gossip/gossip.go:627  missing sentinel gossip; assuming partition in gossip network
W1215 22:42:38.199841 564 gossip/gossip.go:627  missing sentinel gossip; assuming partition in gossip network
W1215 22:42:39.096644 564 gossip/gossip.go:627  missing sentinel gossip; assuming partition in gossip network
W1215 22:42:39.956195 564 gossip/gossip.go:627  missing sentinel gossip; assuming partition in gossip network
W1215 22:42:40.886889 564 gossip/gossip.go:627  missing sentinel gossip; assuming partition in gossip network
--- FAIL: TestClientDisconnectLoopback (10.09s)
	testing.go:157: gossip/client_test.go:266: condition failed to evaluate within 10s: local client still connected to itself
=== RUN   TestClientDisconnectRedundant
I1215 22:42:41.413824 564 gossip/gossip.go:189  setting node descriptor node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:0" > attrs:<> 
I1215 22:42:41.413969 564 gossip/gossip.go:189  setting node descriptor node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:0" > attrs:<> 
W1215 22:42:41.414254 564 rpc/server.go:290  rpc: write response failed: write tcp 127.0.0.1:55256->127.0.0.1:48120: use of closed network connection
I1215 22:42:41.415266 564 gossip/gossip.go:701  starting client to 127.0.0.1:56407
I1215 22:42:41.415293 564 gossip/gossip.go:701  starting client to 127.0.0.1:36151
I1215 22:42:41.416764 564 gossip/server.go:128  received map[node:1:value:<raw_bytes:"\010\001\022\022\n\003tcp\022\013127.0.0.1:0\032\000" checksum:1763343447 timestamp:<wall_time:1450219361413842184 logical:0 > tag:BYTES > orig_stamp:1450219361413847935 ttl_stamp:9223372036854775807 node_id:1  local-key:value:<checksum:1462288794 timestamp:<wall_time:1450219361415307754 logical:0 > tag:BYTES > orig_stamp:1450219361415312011 ttl_stamp:1450219362415307754 node_id:1 ] from node 1 (2 fresh)
I1215 22:42:41.416931 564 gossip/client.go:157  sent 2 info(s) to node 2
I1215 22:42:41.417060 564 gossip/client.go:151  received map[node:2:value:<raw_bytes:"\010\002\022\022\n\003tcp\022\013127.0.0.1:0\032\000" checksum:705884985 timestamp:<wall_time:1450219361413980739 logical:0 > tag:BYTES > orig_stamp:1450219361413983627 ttl_stamp:9223372036854775807 node_id:2 ] from node 2 (1 fresh)
--
I1215 22:42:41.539910 564 gossip/client.go:76  closing client to node 1 (127.0.0.1:38350): gossip/client.go:247: client closed
I1215 22:42:41.539941 564 gossip/client.go:157  sent 1 info(s) to node 4
I1215 22:42:41.539980 564 gossip/client.go:76  closing client to node 4 (127.0.0.1:60632): gossip/client.go:247: client closed
I1215 22:42:41.540301 564 gossip/client.go:151  received map[127.0.0.1:48264:value:<raw_bytes:"\000\000\000\000\000\000\000\004" checksum:2488420480 timestamp:<wall_time:1450219361517897329 logical:0 > tag:BYTES > orig_stamp:1450219361517900460 ttl_stamp:1450222961517897329 hops:2 node_id:8 peer_id:2  127.0.0.1:39009:value:<raw_bytes:"\000\000\000\000\000\000\000\005" checksum:2713591024 timestamp:<wall_time:1450219361524402226 logical:0 > tag:BYTES > orig_stamp:1450219361524403632 ttl_stamp:1450222961524402226 hops:1 node_id:2 peer_id:2  sentinel:value:<raw_bytes:"\000\000\000\000\000\000\000\005" checksum:470669761 timestamp:<wall_time:1450219361524384445 logical:0 > tag:BYTES > orig_stamp:1450219361524386732 ttl_stamp:1450222961524384445 node_id:1  127.0.0.1:38350:value:<raw_bytes:"\000\000\000\000\000\000\000\005" checksum:3916647297 timestamp:<wall_time:1450219361524393004 logical:0 > tag:BYTES > orig_stamp:1450219361524394670 ttl_stamp:1450222961524393004 node_id:1  127.0.0.1:53452:value:<raw_bytes:"\000\000\000\000\000\000\000\004" checksum:1221475777 timestamp:<wall_time:1450219361519163847 logical:0 > tag:BYTES > orig_stamp:1450219361519167146 ttl_stamp:1450222961519163847 hops:1 node_id:10 peer_id:10 ] from node 1 (5 fresh)
--- PASS: TestConvergence (0.05s)
FAIL
FAIL	github.com/cockroachdb/cockroach/gossip	10.280s
=== RUN   TestParseResolverSpec
--- PASS: TestParseResolverSpec (0.00s)
=== RUN   TestGetAddress
--- PASS: TestGetAddress (0.00s)
PASS
ok  	github.com/cockroachdb/cockroach/gossip/resolver	0.008s
?   	github.com/cockroachdb/cockroach/gossip/simulation	[no test files]
=== RUN   TestKeySorting
--- PASS: TestKeySorting (0.00s)
=== RUN   TestMakeKey

```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client_test.go
+++ b/gossip/client_test.go
@@ -251,10 +251,7 @@ func TestClientDisconnectLoopback(t *testing.T) {
 	// startClient requires locks are held, so acquire here.
 	local.mu.Lock()
 	lAddr := local.is.NodeAddr
-	lclock := hlc.NewClock(hlc.UnixNano)
-	rpcContext := rpc.NewContext(&base.Context{Insecure: true}, lclock, stopper)
-	rpcContext.DisableCache = true
-	local.startClient(lAddr, rpcContext, stopper)
+	local.startClient(lAddr, stopper)
 	local.mu.Unlock()
 	local.manage(stopper)
 	util.SucceedsWithin(t, 10*time.Second, func() error {
@@ -279,11 +276,8 @@ func TestClientDisconnectRedundant(t *testing.T) {
 
 	rAddr := remote.is.NodeAddr
 	lAddr := local.is.NodeAddr
-	lclock := hlc.NewClock(hlc.UnixNano)
-	rpcContext := rpc.NewContext(&base.Context{Insecure: true}, lclock, stopper)
-	rpcContext.DisableCache = true
-	local.startClient(rAddr, rpcContext, stopper)
-	remote.startClient(lAddr, rpcContext, stopper)
+	local.startClient(rAddr, stopper)
+	remote.startClient(lAddr, stopper)
 	local.mu.Unlock()
 	remote.mu.Unlock()
 	local.manage(stopper)
@@ -316,14 +310,11 @@ func TestClientDisallowMultipleConns(t *testing.T) {
 	local.mu.Lock()
 	remote.mu.Lock()
 	rAddr := remote.is.NodeAddr
-	lclock := hlc.NewClock(hlc.UnixNano)
-	rpcContext := rpc.NewContext(&base.Context{Insecure: true}, lclock, stopper)
-	rpcContext.DisableCache = true
 	// Start two clients from local to remote. RPC client cache is
 	// disabled via the context, so we'll start two different outgoing
 	// connections.
-	local.startClient(rAddr, rpcContext, stopper)
-	local.startClient(rAddr, rpcContext, stopper)
+	local.startClient(rAddr, stopper)
+	local.startClient(rAddr, stopper)
 	local.mu.Unlock()
 	remote.mu.Unlock()
 	local.manage(stopper)

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -110,8 +110,7 @@ var (
 type Gossip struct {
 	Connected     chan struct{}       // Closed upon initial connection
 	hasConnected  bool                // Set first time network is connected
-	RPCContext    *rpc.Context        // The context required for RPC
-	bsRPCContext  *rpc.Context        // Context for bootstrap RPCs
+	rpcContext    *rpc.Context        // The context required for RPC
 	*server                           // Embedded gossip RPC server
 	outgoing      nodeSet             // Set of outgoing client node IDs
 	bootstrapping map[string]struct{} // Set of active bootstrap clients
@@ -140,7 +139,6 @@ type Gossip struct {
 func New(rpcContext *rpc.Context, resolvers []resolver.Resolver) *Gossip {
 	g := &Gossip{
 		Connected:     make(chan struct{}),
-		RPCContext:    rpcContext,
 		server:        newServer(),
 		outgoing:      makeNodeSet(1),
 		bootstrapping: map[string]struct{}{},
@@ -150,13 +148,15 @@ func New(rpcContext *rpc.Context, resolvers []resolver.Resolver) *Gossip {
 		resolverIdx:   len(resolvers) - 1,
 		resolvers:     resolvers,
 	}
-	// Create the bootstrapping RPC context. This context doesn't
-	// measure clock offsets and doesn't cache clients because bootstrap
-	// connections may go through a load balancer.
+	// The gossip RPC context doesn't measure clock offsets, isn't
+	// shared with the other RPC clients which the node may be using,
+	// and disables reconnects to make it possible to know for certain
+	// which other nodes in the cluster are incoming via gossip.
 	if rpcContext != nil {
-		g.bsRPCContext = rpcContext.Copy()
-		g.bsRPCContext.DisableCache = true
-		g.bsRPCContext.RemoteClocks = nil
+		g.rpcContext = rpcContext.Copy()
+		g.rpcContext.DisableCache = true
+		g.rpcContext.DisableReconnects = true
+		g.rpcContext.RemoteClocks = nil
 	}
 
 	// Add ourselves as a SystemConfig watcher.
@@ -519,7 +519,7 @@ func (g *Gossip) bootstrap(stopper *stop.Stopper) {
 			if !haveClients || !haveSentinel {
 				// Try to get another bootstrap address from the resolvers.
 				if addr := g.getNextBootstrapAddress(); addr != nil {
-					g.startClient(addr, g.bsRPCContext, stopper)
+					g.startClient(addr, stopper)
 				}
 			}
 			g.mu.Unlock()
@@ -591,7 +591,7 @@ func (g *Gossip) doCheckNetwork(stopper *stop.Stopper) {
 		if nodeAddr, err := g.getNodeIDAddressLocked(nodeID); err != nil {
 			log.Errorf("node %d: %s", nodeID, err)
 		} else {
-			g.startClient(nodeAddr, g.RPCContext, stopper)
+			g.startClient(nodeAddr, stopper)
 		}
 	} else if !g.outgoing.hasSpace() {
 		// Otherwise, find least useful peer and close it. Make sure
@@ -612,7 +612,7 @@ func (g *Gossip) doDisconnected(stopper *stop.Stopper, c *client) {
 
 	// If the client was disconnected with a forwarding address, connect now.
 	if c.forwardAddr != nil {
-		g.startClient(c.forwardAddr, g.RPCContext, stopper)
+		g.startClient(c.forwardAddr, stopper)
 	}
 	g.maybeSignalStalledLocked()
 }
@@ -697,13 +697,13 @@ func (g *Gossip) checkHasConnected() {
 // startClient launches a new client connected to remote address.
 // The client is added to the outgoing address set and launched in
 // a goroutine.
-func (g *Gossip) startClient(addr net.Addr, context *rpc.Context, stopper *stop.Stopper) {
+func (g *Gossip) startClient(addr net.Addr, stopper *stop.Stopper) {
 	log.Infof("starting client to %s", addr)
 	c := newClient(addr)
 	g.clientsMu.Lock()
 	g.clients = append(g.clients, c)
 	g.clientsMu.Unlock()
-	c.start(g, g.disconnected, context, stopper)
+	c.start(g, g.disconnected, g.rpcContext, stopper)
 }
 
 // closeClient finds and removes a client from the clients slice.

--- a/gossip/server.go
+++ b/gossip/server.go
@@ -88,6 +88,10 @@ func (s *server) Gossip(argsI proto.Message) (proto.Message, error) {
 	if err != nil {
 		return nil, util.Errorf("local addr %s could not be converted to net.Addr: %s", args.LAddr, err)
 	}
+	// Verify that the client connection is valid and hasn't been closed.
+	if _, ok := s.lAddrMap[lAddr.String()]; !ok {
+		return nil, util.Errorf("connection already closed; ignoring gossip")
+	}
 
 	reply.NodeID = s.is.NodeID
 
@@ -222,6 +226,7 @@ func (s *server) start(rpcServer *rpc.Server, addr net.Addr, stopper *stop.Stopp
 	if err := rpcServer.Register("Gossip.Gossip", s.Gossip, &Request{}); err != nil {
 		log.Fatalf("unable to register gossip service with RPC server: %s", err)
 	}
+	rpcServer.AddOpenCallback(s.onOpen)
 	rpcServer.AddCloseCallback(s.onClose)
 
 	updateCallback := func(_ string, _ roachpb.Value) {
@@ -247,6 +252,21 @@ func (s *server) stop(unregister func()) {
 	unregister()
 	s.closed = true
 	s.ready.Broadcast() // wake up clients
+}
+
+// onOpen is invoked by the rpcServer each time a client connects.
+// Add a placeholder value for the lAddrMap. This prevents races
+// where inflight RPCs re-add gossip clients to the incoming map
+// even when the incoming connection has been closed out from
+// under them.
+//
+// TODO(spencer): remove use of onOpen callback once gossip is using
+//   streaming with gRPC.
+func (s *server) onOpen(conn net.Conn) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	remoteAddr := conn.RemoteAddr().String()
+	s.lAddrMap[remoteAddr] = clientInfo{}
 }
 
 // onClose is invoked by the rpcServer each time a connected client

--- a/kv/dist_sender.go
+++ b/kv/dist_sender.go
@@ -115,6 +115,7 @@ type DistSender struct {
 	// RPCSend is used to send RPC calls and defaults to rpc.Send
 	// outside of tests.
 	rpcSend         rpcSendFn
+	rpcContext      *rpc.Context
 	rpcRetryOptions retry.Options
 }
 
@@ -142,6 +143,7 @@ type DistSenderContext struct {
 	// The RPC dispatcher. Defaults to rpc.Send but can be changed here
 	// for testing purposes.
 	RPCSend           rpcSendFn
+	RPCContext        *rpc.Context
 	RangeDescriptorDB RangeDescriptorDB
 }
 
@@ -184,6 +186,9 @@ func NewDistSender(ctx *DistSenderContext, gossip *gossip.Gossip) *DistSender {
 	ds.rpcSend = rpc.Send
 	if ctx.RPCSend != nil {
 		ds.rpcSend = ctx.RPCSend
+	}
+	if ctx.RPCContext != nil {
+		ds.rpcContext = ctx.RPCContext
 	}
 	ds.rpcRetryOptions = defaultRPCRetryOptions
 	if ctx.RPCRetryOptions != nil {
@@ -359,8 +364,7 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, rangeID roachpb.RangeID, repl
 	}
 
 	const method = "Node.Batch"
-	replies, err := ds.rpcSend(rpcOpts, method, addrs, getArgs, getReply,
-		ds.gossip.RPCContext)
+	replies, err := ds.rpcSend(rpcOpts, method, addrs, getArgs, getReply, ds.rpcContext)
 	if err != nil {
 		return nil, err
 	}

--- a/kv/dist_sender_server_test.go
+++ b/kv/dist_sender_server_test.go
@@ -246,7 +246,7 @@ func TestMultiRangeScanReverseScanInconsistent(t *testing.T) {
 	// the first key to verify it's used to read only key "a".
 	manual := hlc.NewManualClock(ts[1].UnixNano() - 1)
 	clock := hlc.NewClock(manual.UnixNano)
-	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: clock}, s.Gossip())
+	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: clock, RPCContext: s.RPCContext()}, s.Gossip())
 
 	// Scan.
 	sa := roachpb.NewScan(roachpb.Key("a"), roachpb.Key("c"), 0).(*roachpb.ScanRequest)

--- a/rpc/client.go
+++ b/rpc/client.go
@@ -82,16 +82,16 @@ type Client struct {
 	// It signals the end of the heartbeat run loop. When the run loop
 	// exits, it `close()`es `Closed`, which signals to the outside
 	// that the client has indeed stopped.
-	closer, Closed chan struct{}
-	conn           unsafe.Pointer // holds a `internalConn`
-	healthy        atomic.Value   // holds a `chan struct{}` exposed in `Healthy`
-	healthWaitTime time.Time
-	healthReceived chan struct{}
-	tlsConfig      *tls.Config
-
-	clock        *hlc.Clock
-	remoteClocks *RemoteClockMonitor
-	remoteOffset RemoteOffset
+	closer, Closed    chan struct{}
+	conn              unsafe.Pointer // holds a `internalConn`
+	healthy           atomic.Value   // holds a `chan struct{}` exposed in `Healthy`
+	healthWaitTime    time.Time
+	healthReceived    chan struct{}
+	tlsConfig         *tls.Config
+	disableReconnects bool
+	clock             *hlc.Clock
+	remoteClocks      *RemoteClockMonitor
+	remoteOffset      RemoteOffset
 }
 
 // NewClient returns a client RPC stub for the specified address
@@ -124,13 +124,14 @@ func NewClient(addr net.Addr, context *Context) *Client {
 	}
 
 	c := &Client{
-		closer:       make(chan struct{}),
-		Closed:       make(chan struct{}),
-		key:          key,
-		addr:         unresolvedAddr,
-		tlsConfig:    tlsConfig,
-		clock:        context.localClock,
-		remoteClocks: context.RemoteClocks,
+		closer:            make(chan struct{}),
+		Closed:            make(chan struct{}),
+		key:               key,
+		addr:              unresolvedAddr,
+		tlsConfig:         tlsConfig,
+		disableReconnects: context.DisableReconnects,
+		clock:             context.localClock,
+		remoteClocks:      context.RemoteClocks,
 	}
 
 	c.healthy.Store(make(chan struct{}))
@@ -291,6 +292,10 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 
 			// Reconnect on failure.
 			if err != nil {
+				// If reconnects are disabled, and we already have a failed connection, return now.
+				if c.disableReconnects && c.internalConn() != nil {
+					return
+				}
 				if err = c.connect(); err != nil {
 					setUnhealthy()
 					log.Warning(err)

--- a/rpc/rpc.go
+++ b/rpc/rpc.go
@@ -13,11 +13,12 @@ type Context struct {
 	// Embed the base context.
 	base.Context
 
-	localClock   *hlc.Clock
-	Stopper      *stop.Stopper
-	RemoteClocks *RemoteClockMonitor
-	DisableCache bool // Disable client cache when calling NewClient()
-	HealthWait   time.Duration
+	localClock        *hlc.Clock
+	Stopper           *stop.Stopper
+	RemoteClocks      *RemoteClockMonitor
+	DisableCache      bool // Disable client cache when calling NewClient()
+	DisableReconnects bool // Disable client reconnects
+	HealthWait        time.Duration
 }
 
 // NewContext creates an rpc Context with the supplied values.

--- a/rpc/server.go
+++ b/rpc/server.go
@@ -61,6 +61,7 @@ type Server struct {
 
 	mu             sync.RWMutex
 	activeConns    map[net.Conn]struct{}
+	openCallbacks  []func(conn net.Conn)
 	closeCallbacks []func(conn net.Conn)
 	methods        map[string]method
 }
@@ -134,6 +135,22 @@ func (s *Server) RegisterAsync(name string, public bool,
 	return nil
 }
 
+// AddOpenCallback adds a callback to the openCallbacks slice to
+// be invoked when a connection is opened.
+func (s *Server) AddOpenCallback(cb func(conn net.Conn)) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.openCallbacks = append(s.openCallbacks, cb)
+}
+
+func (s *Server) runOpenCallbacks(conn net.Conn) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for _, cb := range s.openCallbacks {
+		cb(conn)
+	}
+}
+
 // AddCloseCallback adds a callback to the closeCallbacks slice to
 // be invoked when a connection is closed.
 func (s *Server) AddCloseCallback(cb func(conn net.Conn)) {
@@ -181,6 +198,9 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		http.Error(w, err.Error(), http.StatusUnauthorized)
 		return
 	}
+
+	// Run open callbacks.
+	s.runOpenCallbacks(conn)
 
 	codec := codec.NewServerCodec(conn)
 	responses := make(chan serverResponse)

--- a/server/node_test.go
+++ b/server/node_test.go
@@ -75,7 +75,7 @@ func createTestNode(addr net.Addr, engines []engine.Engine, gossipBS net.Addr, t
 		g.Start(rpcServer, ln.Addr(), stopper)
 	}
 	ctx.Gossip = g
-	sender := kv.NewDistSender(&kv.DistSenderContext{Clock: ctx.Clock}, g)
+	sender := kv.NewDistSender(&kv.DistSenderContext{Clock: ctx.Clock, RPCContext: nodeRPCContext}, g)
 	ctx.DB = client.NewDB(sender)
 	// TODO(bdarnell): arrange to have the transport closed.
 	// (or attach LocalRPCTransport.Close to the stopper)

--- a/server/server.go
+++ b/server/server.go
@@ -67,6 +67,7 @@ type Server struct {
 
 	mux           *http.ServeMux
 	clock         *hlc.Clock
+	rpcContext    *crpc.Context
 	rpc           *crpc.Server
 	gossip        *gossip.Gossip
 	storePool     *storage.StorePool
@@ -113,25 +114,25 @@ func NewServer(ctx *Context, stopper *stop.Stopper) (*Server, error) {
 	}
 	s.clock.SetMaxOffset(ctx.MaxOffset)
 
-	rpcContext := crpc.NewContext(&ctx.Context, s.clock, stopper)
+	s.rpcContext = crpc.NewContext(&ctx.Context, s.clock, stopper)
 	stopper.RunWorker(func() {
-		rpcContext.RemoteClocks.MonitorRemoteOffsets(stopper)
+		s.rpcContext.RemoteClocks.MonitorRemoteOffsets(stopper)
 	})
 
-	s.rpc = crpc.NewServer(rpcContext)
+	s.rpc = crpc.NewServer(s.rpcContext)
 
-	s.gossip = gossip.New(rpcContext, s.ctx.GossipBootstrapResolvers)
+	s.gossip = gossip.New(s.rpcContext, s.ctx.GossipBootstrapResolvers)
 	s.storePool = storage.NewStorePool(s.gossip, s.clock, ctx.TimeUntilStoreDead, stopper)
 
 	feed := util.NewFeed(stopper)
 	tracer := tracer.NewTracer(feed, ctx.Addr)
 
-	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.clock}, s.gossip)
+	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.clock, RPCContext: s.rpcContext}, s.gossip)
 	sender := kv.NewTxnCoordSender(ds, s.clock, ctx.Linearizable, tracer, s.stopper)
 	s.db = client.NewDB(sender)
 
 	var err error
-	s.raftTransport, err = newRPCTransport(s.gossip, s.rpc, rpcContext)
+	s.raftTransport, err = newRPCTransport(s.gossip, s.rpc, s.rpcContext)
 	if err != nil {
 		return nil, err
 	}

--- a/server/server_test.go
+++ b/server/server_test.go
@@ -299,7 +299,7 @@ func TestMultiRangeScanDeleteRange(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	s := StartTestServer(t)
 	defer s.Stop()
-	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.Clock()}, s.Gossip())
+	ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.Clock(), RPCContext: s.RPCContext()}, s.Gossip())
 	tds := kv.NewTxnCoordSender(ds, s.Clock(), testContext.Linearizable, nil, s.stopper)
 
 	if err := s.node.ctx.DB.AdminSplit("m"); err != nil {
@@ -390,7 +390,7 @@ func TestMultiRangeScanWithMaxResults(t *testing.T) {
 
 	for i, tc := range testCases {
 		s := StartTestServer(t)
-		ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.Clock()}, s.Gossip())
+		ds := kv.NewDistSender(&kv.DistSenderContext{Clock: s.Clock(), RPCContext: s.RPCContext()}, s.Gossip())
 		tds := kv.NewTxnCoordSender(ds, s.Clock(), testContext.Linearizable, nil, s.stopper)
 
 		for _, sk := range tc.splitKeys {

--- a/server/testserver.go
+++ b/server/testserver.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/rpc"
 	"github.com/cockroachdb/cockroach/security"
 	"github.com/cockroachdb/cockroach/storage"
 	"github.com/cockroachdb/cockroach/storage/engine"
@@ -114,6 +115,14 @@ func (ts *TestServer) Gossip() *gossip.Gossip {
 func (ts *TestServer) Clock() *hlc.Clock {
 	if ts != nil {
 		return ts.clock
+	}
+	return nil
+}
+
+// RPCContext returns the rpc context used by the TestServer.
+func (ts *TestServer) RPCContext() *rpc.Context {
+	if ts != nil {
+		return ts.rpcContext
 	}
 	return nil
 }

